d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/QuarantineManager.java

```diff
@@ -0,0 +1,391 @@
+      if (healthScore <= RelativeStateUpdater.MIN_HEALTH_SCORE)
+      {
+        // Reset the health score to initial recovery health score if health score dropped to 0 before
+        trackerClientState.setHealthScore(INITIAL_RECOVERY_HEALTH_SCORE);
+      }
+      else if (_fastRecoveryEnabled)
+      {
+        // If fast recovery is enabled, we perform fast recovery: double the health score
+        healthScore *= FAST_RECOVERY_FACTOR;
+        trackerClientState.setHealthScore(Math.min(healthScore, RelativeStateUpdater.MAX_HEALTH_SCORE));
+      }
+    }
+    else if (!_fastRecoveryEnabled
+        || !trackerClientState.isUnhealthy()
+        || trackerClientState.getHealthScore() > FAST_RECOVERY_HEALTH_SCORE_THRESHOLD)
+    {
+      /**
+       * Remove the client from the map if:
+       * 1. fast recovery is not enabled OR
+       * 2. the client is not unhealthy any more OR
+       * 3. The health score is beyond 0.5, we will let it perform normal recovery
+       */
+      recoverySet.remove(trackerClient);
+    }
+  }
+
+  /**
+   * To put a TrackerClient into quarantine, it needs to meet all the following criteria:
+   * 1. its health score is less than or equal to the threshold (0.0).
+   * 2. The call state in current interval is becoming worse, eg the latency or error rate is higher than the threshold.
+   * 3. its clientWeight is greater than 0
+   *    (ClientWeight can be 0 when the server's clientWeight in zookeeper is explicitly set to 0 in order to put the server into standby.
+   *    In this particular case, we should not put the tracker client into the quarantine).
+   * 4. The total clients in the quarantine is less than the pre-configured number max percentage
+   *
+   * @param trackerClient The server to be quarantined
+   * @param trackerClientState The current state of the server
+   * @param serverWeight The weight of the server host specified from Zookeeper
+   * @param quarantineMap A map of current quarantined hosts
+   * @param quarantineHistory The hosts that used to be quarantined
+   * @param trackerClientSize The total number of hosts in the partition
+   * @param quarantineLatency The quarantine latency threshold
+   * @return True if the host is quarantined
+   */
+  private boolean enrollClientInQuarantineMap(TrackerClient trackerClient, TrackerClientState trackerClientState,
+      double serverWeight, Map<TrackerClient, LoadBalancerQuarantine> quarantineMap,
+      Map<TrackerClient, LoadBalancerQuarantine> quarantineHistory, int trackerClientSize, long quarantineLatency)
+  {
+    if (_quarantineEnabled.get())
+    {
+      double healthScore = trackerClientState.getHealthScore();
+
+      if (quarantineMap.containsKey(trackerClient))
+      {
+        // If the client is still in quarantine, keep the points to 0 so no real traffic will be used
+        trackerClientState.setHealthScore(RelativeStateUpdater.MIN_HEALTH_SCORE);
+        return true;
```

Comment:
I thought about this too, I think it will always be 0 as long as it is in quarantine state. I saw this logic in the V3 implementation, so I just didn't bother to reenforce the health score to be 0 again.
I can remove this reset to 0 logic, it should always be 0, and doesn't need to be reset.